tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -1340,35 +1340,46 @@ void Trace::create_replay() {
     }
 }
 
-void EnqueueReadBuffer(CommandQueue& cq, Buffer& buffer, vector<uint32_t>& dst, bool blocking) {
+void EnqueueReadBuffer(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer>> buffer, vector<uint32_t>& dst, bool blocking){
     // TODO(agrebenisan): Move to deprecated
     ZoneScoped;
     tt_metal::detail::DispatchStateCheck(true);
     TT_FATAL(blocking, "Non-blocking EnqueueReadBuffer not yet supported");
-
+    std::reference_wrapper<Buffer> b = std::holds_alternative<std::shared_ptr<Buffer>>(buffer) ? std::ref (*(std::get< std::shared_ptr<Buffer> > ( buffer ))) :
+                                                                                                 std::ref ( std::get<std::reference_wrapper<Buffer>>(buffer) ) ;
     // Only resizing here to keep with the original implementation. Notice how in the void*
     // version of this API, I assume the user mallocs themselves
-    dst.resize(buffer.page_size() * buffer.num_pages() / sizeof(uint32_t));
-    cq.enqueue_read_buffer(buffer, dst.data(), blocking);
+    dst.resize(b.get().page_size() * b.get().num_pages() / sizeof(uint32_t));
+    cq.enqueue_read_buffer(b, dst.data(), blocking);
 }
 
-void EnqueueWriteBuffer(CommandQueue& cq, Buffer& buffer, vector<uint32_t>& src, bool blocking) {
+void EnqueueWriteBuffer(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer>> buffer, vector<uint32_t>& src, bool blocking){
     // TODO(agrebenisan): Move to deprecated
     ZoneScoped;
     tt_metal::detail::DispatchStateCheck(true);
-    cq.enqueue_write_buffer(buffer, src.data(), blocking);
+
+    std::reference_wrapper<Buffer> b = std::holds_alternative<std::shared_ptr<Buffer>>(buffer) ? std::ref (*(std::get< std::shared_ptr<Buffer> > ( buffer ))) :
```

Comment:
At this point, can we not ensure that type b is just Buffer&? I don't believe it needs to be a reference wrapper. I understand why the variant needs to accept a ref wrapper, but I don't believe b needs to be one.

Reply:
updated all 4 functions